editor/src/clj/editor/lsp.clj

```diff
@@ -0,0 +1,440 @@
+  ;; bound-fn only needed for tests to pick up the test system
+  (a/put! lsp (bound-fn check-if-polled-resources-are-modified [{:keys [polled-resources] :as state}]
+                (g/with-auto-evaluation-context evaluation-context
+                  (reduce #(sync-resource-state! %1 %2 evaluation-context) state polled-resources)))))
+
+(defn apply-resource-changes!
+  "Notify the LSP manager about resource sync"
+  [lsp {:keys [removed changed moved]}]
+  ;; bound-fn only needed for tests to pick up the test system
+  (a/put! lsp (bound-fn apply-resource-changes [{:keys [project get-resource-node] :as state}]
+                (let [interesting-added (into []
+                                              (keep (fn [[from to]]
+                                                      (when (interesting-resource? state from)
+                                                        to)))
+                                              moved)
+                      interesting-removed (filterv #(interesting-resource? state %) removed)
+                      interesting-changed (filterv #(interesting-resource? state %) changed)
+                      viewed-moved (filterv #(resource-viewed? state (first %)) moved)]
+                  (g/with-auto-evaluation-context evaluation-context
+                    (as-> state $
+                          (reduce #(sync-resource-state! %1 %2 evaluation-context) $ interesting-added)
+                          (reduce #(sync-resource-state! %1 %2 evaluation-context) $ interesting-removed)
+                          (reduce #(sync-resource-state! %1 %2 evaluation-context) $ interesting-changed)
+                          (reduce (fn [acc [from to]]
+                                    (let [view-node (get-in acc [:viewed-resources from])
+                                          resource-node (get-resource-node project to evaluation-context)]
+                                      (-> acc
+                                          (do-close-view view-node)
+                                          (do-open-view view-node to (g/node-value resource-node :lines evaluation-context)))))
+                                  $ viewed-moved)))))))
+
+(defn get-graph-lsp
+  "Given a project's graph id, return the LSP manager"
+  ([graph-id]
+   (get-graph-lsp (g/now) graph-id))
+  ([basis graph-id]
+   (g/graph-value basis graph-id :lsp)))
+
+(defn get-node-lsp
+  "Given a node id in a project graph, return the LSP manager"
+  ([node]
+   (get-node-lsp (g/now) node))
+  ([basis node]
+   (get-graph-lsp basis (g/node-id->graph-id node))))
+
+(comment
+  ;; Stop all LSP servers
+  (a/put! (g/graph-value 1 :lsp)
+          (set-servers #{}))
+  ;; Start json LSP server (install: npm install -g vscode-json-languageserver)
+  (a/put! (g/graph-value 1 :lsp)
+          (set-servers #{{:languages #{"json" "jsonc"}
+                          :launcher {:command ["/opt/homebrew/bin/vscode-json-languageserver" "--stdio"]}}})))
```

Comment:
I suggest playing around with LSP by installing JSON language and launching it (your installation path might vary).

You can also try running 2 identical language servers at the same time e.g. by specifying different languages:
```clj
(a/put! (g/graph-value 1 :lsp)
        (set-servers #{{:languages #{"json" "jsonc"}
                        :launcher {:command ["/opt/homebrew/bin/vscode-json-languageserver" "--stdio"]}}
                       {:languages #{"json"}
                        :launcher {:command ["/opt/homebrew/bin/vscode-json-languageserver" "--stdio"]}}}))
```
Here is an example JSON that you can play with:
```json
{
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "width": 200,
  "height": {"step": 16},
  "data": {"url": "data/movies.json"},
  "encoding": {
    "y": {
      "field": "Major Genre",
      "type": "nominal",
      "axis": null
    }
  },
  "layer": [{
    "mark": {"type": "bar", "color": "#ddd"},
    "encoding": {
      "x": {
        "aggregate": "mean",
        "field": "IMDB Rating",
        "scale": {"domain": [0, 10]},
        "title": "Mean IMDB Ratings"
      }
    }
    }, {
      "mark": {"type": "text", "align": "left", "x": 5},
      "encoding": {
        "text": {"field": "Major Genre"},
        "detail": {"aggregate": "count"}
      }
      }]
    }
```
Try changing some keys in the data structure and see language servers complain.